waiter/src/waiter/process_request.clj

```diff
@@ -250,15 +250,29 @@
                                        (cid/with-correlation-id
                                          correlation-id
                                          (let [complete-trigger-id (utils/unique-identifier)]
+                                           (log/debug throwable "received from ctrl-chan" message)
                                            (deliver complete-triggered-promise complete-trigger-id)
                                            (when (= complete-trigger-id @complete-triggered-promise)
                                              (log/debug "closing request body as" message)
                                              (counters/dec! request-body-streaming-counter)
+                                             ;; the callback is necessary as there is a data race between aborting the
+                                             ;; request and closing of the body channel triggering a normal complete before
+                                             ;; the abort request gets processed.
                                              (when throwable
-                                               (log/error throwable "unable to stream request bytes, aborting request")
-                                               (async/>!! abort-ch throwable))
+                                               (let [callback (fn complete-request-streaming-abort-result-callback [aborted?]
+                                                                (cid/with-correlation-id
+                                                                  correlation-id
+                                                                  (log/info "request byte stream error, result of aborted request:" aborted?)
+                                                                  (async/close! body-ch)))]
+                                                 (log/error throwable "unable to stream request bytes, aborting request")
+                                                 (async/>!! abort-ch [throwable callback])))
                                              (report-request-size-metrics 0 true)
-                                             (async/close! body-ch))))))]
+                                             (if throwable
+                                               ;; TODO avoid the need for the timeout while sending to abort channel
```

Comment:
What would it take to avoid the timeout?

Reply:
I created https://github.com/twosigma/waiter/pull/884 to show what we would need.